src/common/mfu_daos.c

```diff
@@ -919,112 +1056,281 @@ int daos_cleanup(
+
+        /* allocate and setup src_buf */
+        if (alloc_iov_buf(number, size, recxs, src_iov, src_buf, src_buf_len) != 0) {
+            MFU_LOG(MFU_LOG_ERR, "DAOS failed to allocate source buffer.");
+            goto out_err;
+        }
+
+        bool recx_equal = false;
+        uint64_t total_bytes = sum_uint64_t(number, src_buf_len);
+
+        /* fetch recx values from source */
+        rc = daos_obj_fetch(*src_oh, DAOS_TX_NONE, 0, dkey, 1, iod,
+                            &src_sgl, NULL, NULL);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "DAOS object fetch returned with errors "DF_RC, DP_RC(rc));
+            goto out_err;
+        }
+        stats->bytes_read += total_bytes;
+
+        /* Conditionally compare the destination before writing */
+        if (compare_dst) {
+            char*       dst_buf[number];
+            uint64_t    dst_buf_len[number];
+            d_sg_list_t dst_sgl;
+            d_iov_t     dst_iov[number];
+
+            dst_sgl.sg_nr_out = 0;
+            dst_sgl.sg_iovs   = dst_iov;
+            dst_sgl.sg_nr     = number;
+
+            /* allocate and setup dst_buf */
+            if (alloc_iov_buf(number, size, recxs, dst_iov, dst_buf, dst_buf_len) != 0) {
+                MFU_LOG(MFU_LOG_ERR, "DAOS failed to allocate destination buffer.");
+                goto out_err;
+            }
+
+            rc = daos_obj_fetch(*dst_oh, DAOS_TX_NONE, 0, dkey, 1, iod,
+                                &dst_sgl, NULL, NULL);
             if (rc != 0) {
-                MFU_LOG(MFU_LOG_ERR, "DAOS object fetch returned with errors: ", MFU_ERRF,
-	                MFU_ERRP(-MFU_ERR_DAOS));
-                return 1;
+                MFU_LOG(MFU_LOG_ERR, "DAOS object fetch returned with errors "DF_RC, DP_RC(rc));
+                free_iov_buf(number, dst_buf);
+                goto out_err;
+            }
+
+            /* Reset iod values after fetching the destination */
+            (*iod).iod_type  = DAOS_IOD_ARRAY;
+            (*iod).iod_nr    = number;
+            (*iod).iod_recxs = recxs;
+            (*iod).iod_size  = size;
+
+            /* Determine whether all recxs in the dst are equal to the src.
+             * If any recx is different, update all recxs in dst and flag
+             * this akey as different. */
+            if (dst_sgl.sg_nr_out > 0) {
+                stats->bytes_read += total_bytes;
+                recx_equal = true;
+                for (uint32_t i = 0; i < number; i++) {
+                    if (memcmp(src_buf[i], dst_buf[i], src_buf_len[i]) != 0) {
+                        recx_equal = false;
+                        all_dst_equal = false;
+                        break;
+                    }
+                }
             }
+            free_iov_buf(number, dst_buf);
+        }
 
-            /* update fetched recx values and place in destination object */
+        /* Conditionally write to the destination */
+        if (write_dst && !recx_equal) {
```

Comment:
@daltonbohning just trying to make sure I understand the code path for the copy correctly, which uses this function it looks like. So for the case of dcp the compare_dst is false, and the write_dst is on, so this is the update to the destination used in that case?

Reply:
@dsikich Right. for `dcp`, we do _**not**_ compare the src and dst, but we _**do**_ update the dst. Which really just means "read src and write to dst".
The reason for the argument `write_dst` is that I was hoping we could perhaps reuse this function for `dcmp` by passing:
`compare_dst=true, write_dst=false`

Reply:
@dsikich Something we could consider is abstracting each action into a separate function. E.g a function for "compare" and a function for "write". Maybe then the code path would make more sense without all of the clutter in a single function.

Reply:
@daltonbohning yeah, I think that is my main comment about the code. I think it would be a bit easier to understand if the cmp/copy functionality was separate. And I think calling a sync_ function in dcp threw me off a bit at first. I think putting the cmp and sync functionality together makes sense. I know some of this stuff runs together and it makes sense to be able to reuse some of it.

Reply:
@dsikich I agree. There are some other performance improvements that need to be made in this function as well, so I'll include this refactor as part of that (once this initial PR is merged).